File: 2. Add Two Numbers/Solution.cpp

struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    int carry = 0; 
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
      if (!l1 && !l2 && !carry) return nullptr;
      if (!l1 && l2) return l2;
      if (l1 && !l2) return l1;

      int val1 = l1? l1->val : 0; 
      int val2 = l2? l2->val : 0; 
      int value = val1 + val2 + carry; 

      ListNode* node = new ListNode(value % 10);
      carry = value / 10;

      node->next = addTwoNumbers(l1? l1->next : nullptr, l2? l2->next : nullptr);
      return node;
    }
};

File: 2. Add Two Numbers/Solution.py
from typing import Optional


class ListNode(object):
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution(object):
    def addTwoNumbers(
        self, l1: Optional[ListNode], l2: Optional[ListNode]
    ) -> Optional[ListNode]:
        carry = 0
        
        def recurse(one: Optional[ListNode], two: Optional[ListNode]):
            nonlocal carry

            if not one and not two and carry == 0:
                return None

            val1 = one.val if one else 0
            val2 = two.val if two else 0

            total = val1 + val2 + carry
            carry = total // 10

            node = ListNode(total % 10)
            node.next = recurse(one.next if one else None, two.next if two else None)
            return node

        return recurse(l1, l2)
